File: Cpp/hash/q1.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

namespace q1 {
class Solution1 {
 public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> hash;
        for (int i = 0; i < nums.size(); ++i) {
            auto it = hash.find(target - nums[i]);
            if (it != hash.end()) {
                return {it->second, i};
            }
            hash[nums[i]] = i;
        }

        return {};
    }
};
}  // namespace q1

int main() {
    std::cout << "q1" << std::endl;
    return EXIT_SUCCESS;
}

File: Rust/src/hash/q1.rs
use std::collections::HashMap;

trait Q1 {
    fn two_sum(&self, nums: Vec<i32>, target: i32) -> Vec<i32>;
}

struct Solution1;
impl Q1 for Solution1 {
    fn two_sum(&self, nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut hash = HashMap::with_capacity(nums.len());
        for i in 0..nums.len() {
            if let Some(key) = hash.get(&(target - nums[i])) {
                if *key != i {
                    return vec![*key as i32, i as i32];
                }
            }
            hash.insert(nums[i], i);
        }
        panic!("not found")
    }
}

pub fn solution() {
    println!("q1")
}
